README.md
# Mollie-Bank [![Gem Version](https://badge.fury.io/rb/mollie-bank.png)][gemversion] [![Build Status](https://secure.travis-ci.org/manuelvanrijn/mollie-bank.png?branch=master)][travis]

[gemversion]: http://badge.fury.io/rb/mollie-bank
[travis]: http://travis-ci.org/manuelvanrijn/mollie-bank

A simple implementation of the "TM Bank" by [Mollie](http://www.mollie.nl), but without the portforward stuff to test your iDeal transactions localy.

![Mollie Bank](http://manuel.manuelles.nl/images/posts/mollie-bank.png)

## Getting started

### Install

To install the gem you should execute

```
gem install mollie-bank
```

### Running the Mollie Bank

After installation you can simple run:

```
mollie-bank
```

Check if it works by browsing to: [http://localhost:4567/](http://localhost:4567/)

## Howto implement

By default all communication for iDeal transactions is through [https://secure.mollie.nl/xml/ideal](https://secure.mollie.nl/xml/ideal). To use the "Mollie Bank" gem, you have to change this path into [http://localhost:4567/xml/ideal](http://localhost:4567/xml/ideal).

Of course you only want to use this in development mode, so you have to create some code to change this path only when it isn't in production mode.

Check the [Wiki: Implement into existing modules](https://github.com/manuelvanrijn/mollie-bank/wiki/Implement-into-existing-modules) page for implementation of existing frameworks/modules.

## Changelog

A detailed overview of can be found in the [CHANGELOG](https://github.com/manuelvanrijn/mollie-bank/blob/master/CHANGELOG.md).

## Copyright

Copyright © 2012 Manuel van Rijn. See [LICENSE](https://github.com/manuelvanrijn/mollie-bank/blob/master/LICENSE.md) for further details.

[![githalytics.com alpha](https://cruel-carlota.pagodabox.com/c7f118ff9d18b3ec8268969b2c5fff65 "githalytics.com")](http://githalytics.com/manuelvanrijn/mollie-bank)


CHANGELOG.md
# Changelog

- **21 September 2014**: version 0.0.5
  - Fixed issue #6: Allow query strings to be present in report URLs by [@rightfold](https://github.com/rightfold)

- **14 May 2013**: version 0.0.4
  - Fixed issue #5: Support both GET and POST requests for `/xml/ideal`
  - Upgraded dependencies

- **14 May 2013**: version 0.0.3
  - Added coveralls
  - Upgraded dependencies

- **23 February 2012**: version 0.0.2
  - New way for storing transactions
  - Fully tested
  - Documented

- **16 February 2012**: version 0.0.1
  - Initial release


spec/errors_spec.rb
require 'spec_helper'

describe "errors" do
  describe "invalid /xml/ideal?a=" do
    it "should return an error -1" do
      post '/xml/ideal', {
        :a => 'unknown'
      }
      xml = Nokogiri::Slop(last_response.body)
      xml.response.item["type"].should eq "error"

      xml.response.item.errorcode.content.should eq "-1"
      xml.response.item.message.content.should eq "Did not receive a proper input value."
    end
  end
  describe "/xml/ideal?a=fetch" do
    it "should return error -2  if partnerid isn't provided" do
      post '/xml/ideal', {
        :a => 'fetch',
        :description => 'description',
        :reporturl => 'http://example.org/report',
        :returnurl => 'http://example.org/return',
        :amount => '1000',
        :bank_id => '0001'
      }
      last_response.should be_ok
      xml = Nokogiri::Slop(last_response.body)
      xml.response.item["type"].should eq "error"

      xml.response.item.errorcode.content.should eq "-2"
      xml.response.item.message.content.should eq "A fetch was issued without specification of 'partnerid'."
    end
    it "should return error -7  if description isn't provided" do
      post '/xml/ideal', {
        :a => 'fetch',
        :partnerid => '1234',
        :reporturl => 'http://example.org/report',
        :returnurl => 'http://example.org/return',
        :amount => '1000',
        :bank_id => '0001'
      }
      last_response.should be_ok
      xml = Nokogiri::Slop(last_response.body)
      xml.response.item["type"].should eq "error"

      xml.response.item.errorcode.content.should eq "-7"
      xml.response.item.message.content.should eq "A fetch was issued without specification of 'description'."
    end
    it "should return error -3  if reporturl isn't provided" do
      post '/xml/ideal', {
        :a => 'fetch',
        :partnerid => '1234',
        :description => 'description',
        :returnurl => 'http://example.org/return',
        :amount => '1000',
        :bank_id => '0001'
      }
      last_response.should be_ok
      xml = Nokogiri::Slop(last_response.body)
      xml.response.item["type"].should eq "error"

      xml.response.item.errorcode.content.should eq "-3"
      xml.response.item.message.content.should eq "A fetch was issued without (proper) specification of 'reporturl'."
    end
    it "should return error -12 if returnurl isn't provided" do
      post '/xml/ideal', {
        :a => 'fetch',
        :partnerid => '1234',
        :description => 'description',
        :reporturl => 'http://example.org/report',
        :amount => '1000',
        :bank_id => '0001'
      }
      last_response.should be_ok
      xml = Nokogiri::Slop(last_response.body)
      xml.response.item["type"].should eq "error"

      xml.response.item.errorcode.content.should eq "-12"
      xml.response.item.message.content.should eq "A fetch was issued without (proper) specification of 'returnurl'."
    end
    it "should return error -4  if amount isn't provided" do
      post '/xml/ideal', {
        :a => 'fetch',
        :partnerid => '1234',
        :description => 'description',
        :reporturl => 'http://example.org/report',
        :returnurl => 'http://example.org/return',
        :bank_id => '0001'
      }
      last_response.should be_ok
      xml = Nokogiri::Slop(last_response.body)
      xml.response.item["type"].should eq "error"

      xml.response.item.errorcode.content.should eq "-4"
      xml.response.item.message.content.should eq "A fetch was issued without specification of 'amount'."
    end
    it "should return error -14 if amount is to low" do
      post '/xml/ideal', {
        :a => 'fetch',
        :partnerid => '1234',
        :description => 'description',
        :reporturl => 'http://example.org/report',
        :returnurl => 'http://example.org/return',
        :amount => '100',
        :bank_id => '0001'
      }
      last_response.should be_ok
      xml = Nokogiri::Slop(last_response.body)
      xml.response.item["type"].should eq "error"

      xml.response.item.errorcode.content.should eq "-14"
      xml.response.item.message.content.should eq "Minimum amount for an ideal transaction is 1,18 EUR."
    end
    it "should return error -6  if bank_id isn't provided" do
      post '/xml/ideal', {
        :a => 'fetch',
        :partnerid => '1234',
        :description => 'description',
        :reporturl => 'http://example.org/report',
        :returnurl => 'http://example.org/return',
        :amount => '1000'
      }
      last_response.should be_ok
      xml = Nokogiri::Slop(last_response.body)
      xml.response.item["type"].should eq "error"

      xml.response.item.errorcode.content.should eq "-6"
      xml.response.item.message.content.should eq "A fetch was issues without specification of a known 'bank_id'."
    end
  end
  describe "/xml/ideal?a=check" do
    it "should return error -11 if partnerid isn't provided" do
      post '/xml/ideal', {
        :a => 'check',
        :transaction_id => '1'
      }
      last_response.should be_ok
      xml = Nokogiri::Slop(last_response.body)
      xml.response.item["type"].should eq "error"

      xml.response.item.errorcode.content.should eq "-11"
      xml.response.item.message.content.should eq "A check was issued without specification of your partner_id."
    end
    it "should return error -8  if transaction_id isn't provided" do
      post '/xml/ideal', {
        :a => 'check',
        :partnerid => '1'
      }
      last_response.should be_ok
      xml = Nokogiri::Slop(last_response.body)
      xml.response.item["type"].should eq "error"

      xml.response.item.errorcode.content.should eq "-8"
      xml.response.item.message.content.should eq "A check was issued without specification of transaction_id."
    end
    it "should return error -10 if transaction_id is unknown" do
      post '/xml/ideal', {
        :a => 'check',
        :partnerid => '1234',
        :transaction_id => '1234'
      }
      last_response.should be_ok
      xml = Nokogiri::Slop(last_response.body)
      xml.response.item["type"].should eq "error"

      xml.response.item.errorcode.content.should eq "-10"
      xml.response.item.message.content.should eq "This is an unknown order."
    end
  end
end


spec/spec_helper.rb
require 'simplecov'
SimpleCov.formatter = SimpleCov::Formatter::MultiFormatter[
  SimpleCov::Formatter::HTMLFormatter,
]
SimpleCov.start

# Load the Sinatra app
require 'mollie-bank'

require 'rspec'
require 'webrat'
require 'rack/test'
require 'nokogiri'

RSpec.configure do |conf|
  conf.include Rack::Test::Methods
  conf.include Webrat::Methods
  conf.include Webrat::Matchers
end

def app
  MollieBank::Application
end


spec/backend_spec.rb
require 'spec_helper'

describe "backend" do
  it "should support GET requests to the /xml/ideal endpoint" do
    get '/xml/ideal', {
      :a => 'banklist'
    }
    last_response.should be_ok
  end
  it "should return the Mollie Bank when requesting /xml/ideal?a=banklist" do
    post '/xml/ideal', {
      :a => 'banklist'
    }

    last_response.should be_ok
    xml = Nokogiri::Slop(last_response.body)
    xml.response.bank.bank_id.content.should eq "0001"
    xml.response.bank.bank_name.content.should eq "Mollie Bank"
  end
  it "should also return the Mollie bank when requesting /xml/ideal/?a=banklist" do
    post '/xml/ideal/', {
      :a => 'banklist'
    }

    last_response.should be_ok
    xml = Nokogiri::Slop(last_response.body)
    xml.response.bank.bank_id.content.should eq "0001"
    xml.response.bank.bank_name.content.should eq "Mollie Bank"
  end
  it "should return the correct xml for /xml/ideal?a=fetch" do
    post '/xml/ideal', {
      :a => 'fetch',
      :partnerid => '1234',
      :description => 'description',
      :reporturl => 'http://example.org/report',
      :returnurl => 'http://example.org/return',
      :amount => '1000',
      :bank_id => '0001'
    }
    last_response.should be_ok

    xml = Nokogiri::Slop(last_response.body)
    xml.response.order.transaction_id.content.should_not be_blank
    xml.response.order.amount.content.should eq "1000"
    xml.response.order.currency.content.should eq "EUR"
    xml.response.order.URL.content.should contain "http://example.org:4567/ideal?transaction_id="
    xml.response.order.message.content.should eq "Your iDEAL-payment has successfully been setup. Your customer should visit the given URL to make the payment"
  end
  it "should return the true if paid" do
    MollieBank::Storage.set('987654', {
      :paid => true
    })
    post '/xml/ideal', {
      :a => 'check',
      :partnerid => '1234',
      :transaction_id => '987654'
    }

    last_response.should be_ok
    xml = Nokogiri::Slop(last_response.body)
    xml.response.order.payed.content.should eq "true"
  end
  it "should return the false not paid" do
    MollieBank::Storage.set('987654', {
      :paid => false
    })
    post '/xml/ideal', {
      :a => 'check',
      :partnerid => '1234',
      :transaction_id => '987654'
    }

    last_response.should be_ok
    xml = Nokogiri::Slop(last_response.body)
    xml.response.order.payed.content.should eq "false"
  end
end


lib/mollie-bank/application/app.rb
# MollieBank Module
module MollieBank
  # The Sinatra Application
  class Application < Sinatra::Base
    register Sinatra::MultiRoute

    set :static, true
    set :public_folder, File.expand_path('..', __FILE__)

    set :views, File.expand_path('../views/', __FILE__)
    set :haml, { :format => :html5 }

    set :storage, MollieBank::Storage

    def self.get_or_post(path, opts={}, &block)
      get(path, opts, &block)
      post(path, opts, &block)
    end

    # Displays a information page of the Mollie-Bank gem
    get '/' do
      haml :info
    end

    # Displays a bank page for finishing the transaction
    get '/ideal' do
      transaction_id = params[:transaction_id]
      description = params[:description]
      reporturl = params[:reporturl]
      returnurl = params[:returnurl]
      amount = params[:amount]

      if transaction_id.nil? or description.nil? or reporturl.nil? or returnurl.nil? or amount.nil?
        haml :html_error, :locals => { :message => "To few params have been supplied (expected to retrieve 'transaction_id', 'description', 'reporturl', 'returnurl' and 'amount')" }
      else
        cent_amount = amount
        int, frac = ("%.2f" % (amount.to_f/100)).split('.')
        amount = "#{int},#{frac}"

        transaction = settings.storage.get(transaction_id)
        transaction[:reporturl] = reporturl
        transaction[:returnurl] = returnurl
        transaction[:amount] = cent_amount

        settings.storage.set(transaction_id, transaction)

        url_path = request.url.split('/ideal?transaction_id=')[0]

        haml :bank_page, :locals => {
          :transaction_id => transaction_id,
          :amount => amount,
          :reporturl => reporturl,
          :returnurl => returnurl,
          :description => description,
          :url_path => url_path
        }
      end
    end

    # Background page that sends the paid = true or paid = false to the report url
    get '/payment' do
      transaction_id = params[:transaction_id]
      paid = params[:paid] == "true" ? true : false

      if transaction_id.nil? or paid.nil?
        haml :html_error, :locals => { :message => "To few params have been supplied" }
      else
        transaction = settings.storage.get(transaction_id)
        reporturl = transaction[:reporturl]
        returnurl = transaction[:returnurl]
        transaction[:paid] = paid

        settings.storage.set(transaction_id, transaction)

        begin
          reporturl = URI("#{reporturl}#{reporturl.include?('?') ? '&' : '?'}transaction_id=#{transaction_id}")
          Net::HTTP.get(reporturl)
        rescue
        end

        redirect_url = "#{returnurl}&transaction_id=#{transaction_id}"
        # if return url already contains a '?' we have to append the transaction_id with a '&'
        redirect_url = "#{returnurl}?transaction_id=#{transaction_id}" if returnurl.split('?').count == 1

        redirect redirect_url
      end
    end

    # Displays the xml depending on the specified action
    #
    # @example
    #   # returns a list of banks
    #   http://localhost:4567/xml/ideal?a=banklist
    #
    # @example
    #   # creates a new order
    #   http://localhost:4567/xml/ideal?a=fetch
    #
    # @example
    #   # checks if a order was paid
    #   http://localhost:4567/xml/ideal?a=check
    get_or_post '/xml/ideal', '/xml/ideal/' do
      content_type 'text/xml'
      case params[:a]
      when "banklist"
        haml :banklist, :layout => false
      when "fetch"
        return error(-2) unless params.has_key?("partnerid")
        return error(-7) unless params.has_key?("description")
        return error(-3) unless params.has_key?("reporturl")
        return error(-12) unless params.has_key?("returnurl")
        return error(-4) unless params.has_key?("amount")
        return error(-14) unless params[:amount].to_i > 118
        return error(-6) unless params.has_key?("bank_id")

        description = params[:description]
        reporturl = params[:reporturl]
        returnurl = params[:returnurl]
        amount = params[:amount]

        transaction_id = UUID.new.generate.gsub('-', '')

        settings.storage.set(transaction_id, {:paid => false})

        #url_path = request.url.split('/xml/ideal')[0]
        url_path = "http://#{request.host}:#{Sinatra::Application.port}"

        haml :fetch, :layout => false, :locals => {
          :transaction_id => transaction_id,
          :amount => amount,
          :reporturl => reporturl,
          :returnurl => returnurl,
          :description => description,
          :url_path => url_path
        }
      when "check"
        return error(-11) unless params.has_key?("partnerid")
        return error(-8) unless params.has_key?("transaction_id")

        transaction_id = params[:transaction_id]
        transaction = settings.storage.get(transaction_id)
        return error(-10) if transaction.nil?

        is_paid = transaction[:paid]
        amount = transaction[:amount]
        transaction[:paid] = false

        settings.storage.set(transaction_id, transaction)

        haml :check, :layout => false, :locals => {
          :transaction_id => transaction_id,
          :amount => amount,
          :is_paid => is_paid
        }
      else
        error(-1)
      end
    end

    # Render the XML error for a specific code
    #
    # @param [int] code
    def error(code)
      # Mollie codes taken from https://www.mollie.nl/support/documentatie/betaaldiensten/ideal/en/
      errors = []
      errors[1] = "Did not receive a proper input value."
      errors[2] = "A fetch was issued without specification of 'partnerid'."
      errors[3] = "A fetch was issued without (proper) specification of 'reporturl'."
      errors[4] = "A fetch was issued without specification of 'amount'."
      errors[5] = "A fetch was issued without specification of 'bank_id'."
      errors[6] = "A fetch was issues without specification of a known 'bank_id'."
      errors[7] = "A fetch was issued without specification of 'description'."
      errors[8] = "A check was issued without specification of transaction_id."
      errors[9] = "Transaction_id contains illegal characters. (Logged as attempt to mangle)."
      errors[10] = "This is an unknown order."
      errors[11] = "A check was issued without specification of your partner_id."
      errors[12] = "A fetch was issued without (proper) specification of 'returnurl'."
      errors[13] = "This amount is only permitted when iDEAL contract is signed and sent to Mollie."
      errors[14] = "Minimum amount for an ideal transaction is 1,18 EUR."
      errors[15] = "A fetch was issued for an account which is not allowed to accept iDEAL payments (yet)."
      errors[16] = "A fetch was issued for an unknown or inactive profile."

      haml :error, :layout => false, :locals => {
        :type => "error",
        :code => code,
        :message => errors[code*-1]
      }
    end
  end
end
